### ISSUE TITLE
Option to "always" accept cert with mismatching server name doesn't cause it to always be accepted

### ISSUE THREAD

**Issue by felixc** [iNPUTmice/Conversations#1904]
#### General information
- **Version:** 1.12.9
- **Device:** Nexus 5X
- **Android Version:** Android 6.0.1 at May 1, 2016 security patch level
- **Server name:** `felixcrux.com` is the domain name, but under the hood it's a Google Apps account, so really `xmpp-server.l.google.com`
- **Server software:** Unknown
- **Installed server modules:** Unknown
- **Conversations source:** Play Store
#### Steps to reproduce
1. Have a domain name with an associated Google Apps account.  For example in my case I have `felixcrux.com`, but with `dig SRV '_xmpp-server._tcp.felixcrux.com'` you can see that I'm using `xmpp-server.l.google.com` as the actual XMPP server.
2. Set up the account in Conversations by just entering your `user@example.com` address/password and letting the app auto-detect that it should talk to Google's servers.
3. Because Google's TLS certs are valid for `google.com`, rather than your `example.com` custom domain, Conversations determines that there's a problem with the certificate. You'll be prompted to acknowledge the issue, and given details about the cert and what names it's valid for. This is a good thing!
4. At the bottom of the prompt, when asked whether to accept the cert with its mismatching server name, select "Always".
5. Observe that at some point later you'll be asked again, ad infinitum, instead of the cert _always_ being accepted.
#### Expected result

The user should be prompted only once about the mismatch between the domain name and the server cert's domain names. If the user selects the option to "always" accept it, the cert should always be accepted in the future without further prompting the user.

**Comment by felixc** [comment]
Upon some further digging, I think the mismatch between my own domain name and `google.com` is a red herring; the problem is really entirely with the cert on Google's side. i.e. the mismatch isn't between `felixcrux.com` and `google.com`, it's between `xmpp-server.l.google.com` and the names presented by the cert on that server.

That doesn't change the nature of the bug report, which is about the fact that the app keeps asking about the name mismatch, but it does make reproducing it easier: there's no need to have a Google Apps account or multiple domains or anything, just a server with a bad cert name should be enough.

Sorry for the mixup.

**Comment by iNPUTmice** [comment]
Related to #1891?

**Comment by felixc** [comment]
Definitely related in that I see the same issue (i.e. the dialog pops up over and over and stacks on top of itself), but I think there are two problems:
1) The dialog spawns ∞ copies of itself on top of each other, instead of stopping at one.
2) The "always" accept option doesn't accept the cert forever. This is distinct because the next time you're asked might be a day later, i.e. there may only ever be one dialog at a time, with no stacking.

Not sure if you want to treat those as separate bugs, or if they're related enough to work with together.

**Comment by iNPUTmice** [comment]
1 is fixed in 1.13.0 (if you sign up for the beta you can already test
this. Otherwise wait until tomorrow or so)

2 are you sure it's the same certificate. Maybe Google is changing it? Have
a look at the checksums.

**Comment by felixc** [comment]
I will note the checksums the next two times I see this happen, and confirm whether or not they are equal.

**Comment by felixc** [comment]
I can confirm that it is the same cert, and the checksums match.

When I go into the Expert Settings and select "Remove certificates", there is one and only one entry for `felixcrux.com`. That's the domain for my account, but it's not the server's domain. Not sure if that's relevant.

**Comment by iNPUTmice** [comment]
Sorry I just can reproduce this with your server.

**Comment by felixc** [comment]
I can try to get adb logs or something, if that might be helpful?

Not sure how to best proceed, since I definitely observe this behaviour reproducibly on my device, and it sounds like maybe the poster in #1891 does too?

Or perhaps it's fixed at `HEAD` as part of other work? I can wait a bit for updates to roll out and see if it still occurs?

**Comment by iNPUTmice** [comment]
No #1891 was about something else. (Which is fixed)

(On an unrelated side note you really don't want to use gmail as your main XMPP server. Gmail dosen't support any of the extensions that make Conversations so great.)

**Comment by felixc** [comment]
I know, but your page on https://account.conversations.im/ doesn't tell me how to actually sign up to give you money :)

(Also all the usual problems about networks effects, actually being able to contact certain people, etc.)

**Comment by iNPUTmice** [comment]
> To create an account your client needs to support inband registration. Depending on the client, you will either be asked for the desired address (e.g., alice@conversations.im), or separately for the user name (e.g., alice) and domain part (conversations.im). A login password must also be specified. If anything else can be configured, the default settings should usually be fine.

For Conversations that means you just have to check the 'Register New Account on Server' box.

After that you will receive further instructions via XMPP message. (We should probably note that last part on our website as well.)

**Comment by felixc** [comment]
Ah, thanks; I figured it might be the case that further details and next steps would be provided via XMPP itself, but yes, it would be nice to know that for sure in advance and to get some ideas of how it works, what payment methods are supported, etc.

The other big issue for me personally is whether there's a way to support it being *hosted* by you, but externally appearing to be on my own domain. I.e. whether my contact JID (if that's still the right terminology?) can be `felixc@felixcrux.com`.

In any case, I'll certainly try it out. Thanks for the info!

**Comment by AltNico** [comment]
I also had the problem that I always had to accept the certificate, although I clicked "always".

I fixed it by setting the right server name in the certificate. With fixing this issue I realized that it also fixes another error: If you are expericening this issue, you also can't use Conversations over VPN (but it did not ask to confirm the certificate; it just did not work). I always had to turn off VPN to use Conversations, while all other apps worked fine.

With setting the right server name, I can now use Conversations over VPN.

BTW: Thank you for your amazing app!
